c/tskit/tables.c

```diff
@@ -12646,6 +12646,133 @@ tsk_table_collection_union(tsk_table_collection_t *self,
+     * so we're safe while we're calling that.
+     */
+    ret = (int) tsk_table_collection_check_integrity(self, 0);
+    if (ret != 0) {
+        goto out;
+    }
+
+    ret = tsk_edge_table_copy(&self->edges, &edges, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    /* Note: we are assuming below that the tables are sorted, so we could save
+     * a bit of time and memory here by truncating part-way through the
+     * edges. */
+    ret = tsk_edge_table_clear(&self->edges);
+    if (ret != 0) {
+        goto out;
+    }
+    for (j = 0; j < (tsk_id_t) edges.num_rows; j++) {
+        tsk_edge_table_get_row_unsafe(&edges, j, &edge);
+        if (node_time[edge.child] < time) {
+            if (time < node_time[edge.parent]) {
+                ret_id = tsk_node_table_add_row(
+                    &self->nodes, 0, time, TSK_NULL, TSK_NULL, NULL, 0);
+                if (ret_id < 0) {
+                    ret = (int) ret_id;
+                    goto out;
+                }
+                edge.parent = ret_id;
+            }
+            ret_id = tsk_edge_table_add_row(&self->edges, edge.left, edge.right,
+                edge.parent, edge.child, edge.metadata, edge.metadata_length);
+            if (ret_id < 0) {
+                ret = (int) ret_id;
+                goto out;
+            }
+        }
+    }
+    /* Calling x_table_free multiple times is safe, so get rid of the
+     * extra edge table memory as soon as we can. */
+    tsk_edge_table_free(&edges);
+
+    ret = tsk_mutation_table_copy(&self->mutations, &mutations, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_mutation_table_clear(&self->mutations);
+    if (ret != 0) {
+        goto out;
+    }
+    for (j = 0; j < (tsk_id_t) mutations.num_rows; j++) {
+        tsk_mutation_table_get_row_unsafe(&mutations, j, &mutation);
+        mutation_time = tsk_is_unknown_time(mutation.time) ? node_time[mutation.node]
+                                                           : mutation.time;
+        if (mutation_time < time) {
+            /* Set the mutation parent to NULL, and recalculate below */
+            ret_id = tsk_mutation_table_add_row(&self->mutations, mutation.site,
+                mutation.node, TSK_NULL, mutation.time, mutation.derived_state,
+                mutation.derived_state_length, mutation.metadata,
+                mutation.metadata_length);
+            if (ret_id < 0) {
+                ret = (int) ret_id;
+                goto out;
+            }
+        }
+    }
+    tsk_mutation_table_free(&mutations);
+
+    ret = tsk_migration_table_copy(&self->migrations, &migrations, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_migration_table_clear(&self->migrations);
+    if (ret != 0) {
+        goto out;
+    }
+    for (j = 0; j < (tsk_id_t) migrations.num_rows; j++) {
+        tsk_migration_table_get_row_unsafe(&migrations, j, &migration);
+        if (migration.time <= time) {
```

Comment:
Above we had `if (mutation.time < time)`, which is I think the right way (`<` not `<=`) because intervals are open on the right?

Reply:
oh never mind, you're right! This is the right way to go.

Reply:
well, actually, no, I take that back. I think that this should be `<` not `<=` because the interpretation of a new inserted node should be "the ancestor at the time just infinitesimally more recent than `t`", not "at `t`". Currently (with `<=`) the same migration would show up in the decapitated tree sequence as well as in the tree sequence with everything `>= t` in it (a method we don't have implemented, but still). But, it should show up in exactly one of those.

Reply:
My reasoning for making it <= for migrations was that mutations were the exception, and the only reason for making mutations < t was because if we left them in we'd violate the "mutations must be younger than the parent of the edge" rule.

However, if we drop the extra nodes (like suggested below) we could make all the conditions <= which would be much simpler I think. (<= is right because we definitely want to include edges that have time[parent] == t, right?)